### PR TITLE
Fluid-like syntax for markers

### LINKED ISSUE #23: Use Fluid-like syntax for the markers
Markers that are currently using the syntax `#FOO#` should now look like `{foo}`.

Additionally, they should support dotted path syntax to fetch nested values, for instance `{foo.bar}`.

### PR DESCRIPTION
**Pull request by Mopolo**
This fixes #23

### CODE REVIEW

File: Classes/Property/Service/MarkerParser.php [hunk 1]
```diff
@@ -30,12 +31,88 @@ class MarkerParser
      */
     public function replaceMarkers($string, array $markers)
     {
+        if (empty($markers)) {
+            return $string;
+        }
+
+        $markers = $this->keyMarkersByName($markers);
+        $matches = $this->matchMarkers($string);
```
Comment (romm) [inline Classes/Property/Service/MarkerParser.php]:
```php
list($identifiers, $variables, $roots) = $this->matchMarkers($string);

if (empty($identifiers)) {
    // ...
}
```

File: Classes/Property/Service/MarkerParser.php [hunk 2]
```diff
@@ -30,12 +31,88 @@ class MarkerParser
...
+    /**
+     * This method will find all markers in the string
+     *
+     * @param $string
+     * @return mixed
+     */
+    private function matchMarkers($string)
+    {
+        preg_match_all(
+            '/{
```
Comment (romm) [inline Classes/Property/Service/MarkerParser.php:88]: Indentation is broken.

File: Classes/Property/Service/MarkerParser.php [hunk 3]
```diff
@@ -30,12 +31,88 @@ class MarkerParser
...
+        }
+
+        return $aux;
+    }
+
+    /**
+     * This method will find all markers in the string
+     *
+     * @param $string
+     * @return mixed
```
Comment (romm) [inline Classes/Property/Service/MarkerParser.php]: Should always return an array. Are we sure the matches are always filled with an array?

**Review by romm — CHANGES_REQUESTED** [review]

File: Classes/Property/Service/MarkerParser.php (hunk 3 above)
Comment (Mopolo) [reply Classes/Property/Service/MarkerParser.php]: Yes I tried and when there's no match `$matches` is still an array with the correct structure.

File: Classes/Property/Service/MarkerParser.php [hunk 4]
```diff
@@ -81,17 +80,17 @@ private function keyMarkersByName(array $markers)
...
     private function matchMarkers($string)
     {
         preg_match_all(
-            '/{
-                        (
-                            ([a-z]+[a-z0-1]*)           # The root variable
-                            (?:\.[a-z]+[a-z0-1]*)*      # The other parts
-                        )
-                    }/xi',
+    '/{
```
Comment (romm) [inline Classes/Property/Service/MarkerParser.php]: Fail. 😆

**Review by romm — CHANGES_REQUESTED** [review]

**Comment by romm** [comment]
One thing that bothers me: the `ObjectAccess` [can throw an exception if a property can not be accessed](https://github.com/TYPO3/TYPO3.CMS/blob/v8.7.9/typo3/sysext/extbase/Classes/Reflection/ObjectAccess.php#L102).

Considering that the markers come from an input that editors can modify, this is more of a runtime exception. I think we should catch it, to prevent a notification to completely crash in this situation.

Maybe a log entry should be created when this exception is caught?

**Comment by romm** [comment]
Also, the documentation should be updated to show the new markers syntax.

**Comment by Mopolo** [comment]
The method `ObjectAccess::getPropertyPath` does the catch and returns `null` if that happens:

https://github.com/TYPO3/TYPO3.CMS/blob/v8.7.9/typo3/sysext/extbase/Classes/Reflection/ObjectAccess.php#L146

**Review by romm — APPROVED** [review]

**Pull request merged**